pkg/corerp/frontend/controller/applications/graph_util.go

```diff
@@ -497,21 +491,55 @@ func connectionsFromAPIData(resource generated.GenericResource) []*corerpv202310
 		data := corerpv20231001preview.ConnectionProperties{}
 		err := toStronglyTypedData(connection, &data)
 		if err == nil {
+			sourceID, _ := findSourceResource(to.String(data.Source), allResources)
+
 			entries = append(entries, &corerpv20231001preview.ApplicationGraphConnection{
-				ID:        data.Source,
-				Direction: &dir,
+				ID:        to.Ptr(sourceID),
+				Direction: to.Ptr(dir),
 			})
 		}
 	}
 
 	// Produce a stable output
 	sort.Slice(entries, func(i, j int) bool {
-		return *entries[i].ID < *entries[j].ID
+		return to.String(entries[i].ID) < to.String(entries[j].ID)
 	})
 
 	return entries
 }
 
+// findSourceResource looks up resource id by using source string by the following steps:
+// 1. Immediately return the resource ID if the source is a valid resource ID.
+// 2. Parse the hostname from source and look up the hostname in the resource list if the source is a valid URL.
+// 3. Otherwise, return the original source string with false boolean value.
```

Comment:
@nithyatsu @rynowak This returns original source string even if source is invalid. So client can still represent invalid source as individual object on graph. Please let me know if you have any concern.

Reply:
I think this is the right decision. I'd rather err on the side of showing more data then returning an error message.

Reply:
I updated the code to return error.